Panic on call stack (maybe?) overflow

As a result of #1206, I was playing with a test factorial function, and it seems that Elvish panics when a math overflow is reached:
```
[~]|> fn fact [n]{ if (== $n 0) { put 1 } else { put (* $n (fact (- $n 1))) } }
[~]|> fact 50
▶ (float64 3.0414093201713376e+64)
[~]|> fact 60
panic: interface conversion: error is *os.SyscallError, not *eval.Exception

goroutine 24161 [running]:
github.com/elves/elvish/pkg/eval.(*pipelineOp).exec.func1(0x152a5a0, 0xc000dca210, 0xc000a44e70, 0xc00057b540, 0xc001030590, 0x203000)
	github.com/elves/elvish/pkg/eval/compile_effect.go:153 +0x152
created by github.com/elves/elvish/pkg/eval.(*pipelineOp).exec
	github.com/elves/elvish/pkg/eval/compile_effect.go:149 +0x225
Exception: elvish exited with 2
```

I don't think overflow is the problem. After all, floats can get almost as big as 1e308, and 60! is far smaller. Also, I get the crash at `fact 62`, but not for smaller numbers.

Is it a stack overflow, perhaps? After all, elvish is not supposed to be tail call optimised. Here is an example that is clearly not due to math overflow.

```
⬥ fn pow [x n]{ if (== $n 0) { put 1 } else { put (* $x (pow $x (- $n 1))) } }
⬥ pow 2 10
⮕ (float64 1024)
⬥ pow 1.05 60
⮕ (float64 18.679185894123)
⬥ pow 1.05 70
panic: interface conversion: error is *os.SyscallError, not *eval.Exception

goroutine 9238 [running]:
github.com/elves/elvish/pkg/eval.(*pipelineOp).exec.func1(0x1528240, 0xc00054b8c0, 0xc00049c2a0, 0xc0000a4800, 0xc0000a8c80, 0x0)
	/Volumes/solid/pkg/gopkg/src/github.com/elves/elvish/pkg/eval/compile_effect.go:150 +0x152
created by github.com/elves/elvish/pkg/eval.(*pipelineOp).exec
	/Volumes/solid/pkg/gopkg/src/github.com/elves/elvish/pkg/eval/compile_effect.go:146 +0x225
Exception: elvish exited with 2
[tty 5], line 1: elvish

```

--- a/pkg/eval/compile_effect.go
+++ b/pkg/eval/compile_effect.go
@@ -139,7 +139,12 @@ func (op *pipelineOp) exec(fm *Frame) error {
 			err := thisOp.exec(newFm)
 			newFm.Close()
 			if err != nil {
-				*thisError = err.(*Exception)
+				switch err := err.(type) {
+				case *Exception:
+					*thisError = err
+				default:
+					*thisError = &Exception{Reason: err, StackTrace: nil}
+				}
 			}
 			wg.Done()
 			if hasChanInput {

